File: frontend/js/features/german-exam/dsh-hv-playback.ts
// DSH Hörverstehen two-play delivery — browser mirror of
// backend/python-ai/app/services/german_exams/dsh_hv_playback.py (the authoritative rule set).
// Both are checked against the SAME scripted vectors (tests/fixtures/dsh-hv-playback-vectors.json).
//
// MPO §10(1)1 / §10(4)1b: the lecture is presented exactly twice; 10 minutes of processing time follow the
// first and 40 minutes the second presentation. There is no third play and no play during a window.
//
// LIMITATION: this state lives in the browser (storage). A learner can clear it, so it protects against
// reloads, stale tabs and double clicks, not against deliberate tampering. A tamper-proof play count needs a
// server-side attempt record (table + endpoint + migration): FUTURE WORK.

export type HvPhase =
  | 'ready'
  | 'playing_1'
  | 'processing_1'
  | 'awaiting_play_2'
  | 'playing_2'
  | 'processing_2'
  | 'submitted'
  | 'expired';
export type HvEvent = 'start_play' | 'playback_ended' | 'tick' | 'submit';

export interface HvPlaybackState {
  schemaVersion: 1;
  attemptId: string;
  phase: HvPhase;
  playsStarted: number;
  playsCompleted: number;
  windowDeadlineMs: number | null;
}

export const HV_MAX_PLAYS = 2;
const PHASES: HvPhase[] = ['ready', 'playing_1', 'processing_1', 'awaiting_play_2', 'playing_2', 'processing_2', 'submitted', 'expired'];
/** Processing windows after presentation 1 and 2 (MPO §10(1)1); the manifest carries the same numbers. */
export const HV_WINDOWS_MS: Record<1 | 2, number> = { 1: 600_000, 2: 2_400_000 };

export class HvPlaybackError extends Error {}

export function newHvState(attemptId: string): HvPlaybackState {
  if (!attemptId) throw new HvPlaybackError('attemptId is required');
  return { schemaVersion: 1, attemptId, phase: 'ready', playsStarted: 0, playsCompleted: 0, windowDeadlineMs: null };
}

function validate(state: HvPlaybackState): void {
  if (!state || state.schemaVersion !== 1 || !PHASES.includes(state.phase)) throw new HvPlaybackError('invalid playback state');
  if (!Number.isInteger(state.playsStarted) || state.playsStarted < 0 || state.playsStarted > HV_MAX_PLAYS) {
    throw new HvPlaybackError('invalid play count');
  }
}

function tickState(state: HvPlaybackState, nowMs: number): HvPlaybackState {
  const deadline = state.windowDeadlineMs;
  if (deadline === null || nowMs < deadline) return state;
  if (state.phase === 'processing_1') return { ...state, phase: 'awaiting_play_2', windowDeadlineMs: null };
  if (state.phase === 'processing_2') return { ...state, phase: 'expired', windowDeadlineMs: null };
  return state;
}

export function advanceHv(state: HvPlaybackState, event: HvEvent | string, nowMs: number): HvPlaybackState {
  validate(state);
  if (!Number.isInteger(nowMs) || nowMs < 0) throw new HvPlaybackError('nowMs must be a non-negative integer');
  const current = tickState(state, nowMs); // a lapsed window always resolves first, whatever the event is
  const phase = current.phase;
  if (event === 'tick') return current;
  if (event === 'start_play') {
    if (phase === 'ready') return { ...current, phase: 'playing_1', playsStarted: 1 };
    if (phase === 'awaiting_play_2') return { ...current, phase: 'playing_2', playsStarted: 2 };
    throw new HvPlaybackError(`cannot start a play in phase ${phase}`);
  }
  if (event === 'playback_ended') {
    if (phase === 'playing_1' || phase === 'playing_2') {
      const n: 1 | 2 = phase === 'playing_1' ? 1 : 2;
      return { ...current, phase: n === 1 ? 'processing_1' : 'processing_2', playsCompleted: n, windowDeadlineMs: nowMs + HV_WINDOWS_MS[n] };
    }
    throw new HvPlaybackError(`no play is running in phase ${phase}`);
  }
  if (event === 'submit') {
    if (phase === 'processing_2') return { ...current, phase: 'submitted', windowDeadlineMs: null };
    throw new HvPlaybackError(`submission opens after the second presentation (phase ${phase})`);
  }
  throw new HvPlaybackError(`unknown event ${String(event)}`);
}

/** After a reload: an interrupted play counts as consumed and its window starts now; it is never re-granted. */
export function recoverHv(state: HvPlaybackState, nowMs: number): HvPlaybackState {
  validate(state);
  let next = state;
  if (next.phase === 'playing_1' || next.phase === 'playing_2') next = advanceHv(next, 'playback_ended', nowMs);
  return advanceHv(next, 'tick', nowMs);
}

export function canStartPlay(state: HvPlaybackState, nowMs: number): boolean {
  try {
    advanceHv(state, 'start_play', nowMs);
    return true;
  } catch {
    return false;
  }
}

export function remainingMs(state: HvPlaybackState, nowMs: number): number | null {
  const deadline = tickState(state, nowMs).windowDeadlineMs;
  return deadline === null ? null : Math.max(0, deadline - nowMs);
}

/**
 * Persisted controller. `beginPlay()` writes the new state to storage BEFORE it returns true, so the audio is
 * only started after the play is on record; a reload (or a second tab) after that can never see play 1 again.
 * With no usable storage it fails closed: no play can be started that could not be recorded.
 */
export class HvPlaybackController {
  private state: HvPlaybackState;
  private readonly key: string;

  constructor(
    private readonly storage: Pick<Storage, 'getItem' | 'setItem'> | undefined,
    attemptId: string,
    private readonly now: () => number = Date.now
  ) {
    this.key = `dsh-hv-playback:${attemptId}`;
    let loaded: HvPlaybackState | null = null;
    try {
      const raw = storage?.getItem(this.key);
      if (raw) loaded = JSON.parse(raw) as HvPlaybackState;
    } catch {
      loaded = null;
    }
    try {
      this.state = loaded && loaded.attemptId === attemptId ? recoverHv(loaded, this.now()) : newHvState(attemptId);
    } catch {
      // A corrupt record must not restore a play: refuse to start anything rather than start over.
      this.state = { ...newHvState(attemptId), phase: 'expired' };
    }
    this.persist();
  }

  private persist(): boolean {
    if (!this.storage) return false;
    try {
      this.storage.setItem(this.key, JSON.stringify(this.state));
      return true;
    } catch {
      return false;
    }
  }

  get snapshot(): HvPlaybackState {
    this.state = tickState(this.state, this.now());
    return this.state;
  }

  beginPlay(): boolean {
    const before = this.state;
    try {
      this.state = advanceHv(this.state, 'start_play', this.now());
    } catch {
      return false;
    }
    if (!this.persist()) {
      this.state = before; // could not be recorded: do not play
      return false;
    }
    return true;
  }

  endPlay(): void {
    this.state = advanceHv(this.state, 'playback_ended', this.now());
    this.persist();
  }

  submit(): void {
    this.state = advanceHv(this.state, 'submit', this.now());
    this.persist();
  }
}

File: backend/python-ai/app/services/german_exams/dsh_hv_playback.py
"""DSH Hörverstehen delivery state machine — the authoritative rules for the two-play flow.

MPO §10(1)1 / §10(4)1b (2025): the lecture is presented exactly TWICE; processing time is 10 minutes
after the first and 40 minutes after the second presentation (the presentation itself does not count).

    ready --start_play--> playing_1 --playback_ended--> processing_1 --(10 min)--> awaiting_play_2
      --start_play--> playing_2 --playback_ended--> processing_2 --submit / (40 min)--> submitted | expired

Design rules:
  * The state is a plain JSON-serialisable dict; `advance()` is a pure function (state, event, now) ->
    new state. The browser mirrors it (frontend/js/features/german-exam/dsh-hv-playback.ts) and both
    are checked against the SAME vectors (tests/fixtures/dsh-hv-playback-vectors.json).
  * `playsStarted` is incremented when a play STARTS and the caller must persist the new state BEFORE
    the audio starts. A reload in the middle of a play therefore cannot restore the play: `recover()`
    treats an interrupted play as consumed.
  * There is no third play and no play while a processing window runs; an illegal event raises.

LIMITATION (documented, not hidden): a client-held state can be cleared by the user. Making the play
count tamper-proof needs a server-side attempt record (table + endpoint + migration), which is
FUTURE WORK. This module is the rule set that server record will reuse.
"""

from __future__ import annotations

from typing import Any

from .registry import get_part
from .shared import GermanExamProfileError

SCHEMA_VERSION = 1
MAX_PLAYS = 2
PHASES = ("ready", "playing_1", "processing_1", "awaiting_play_2", "playing_2", "processing_2", "submitted", "expired")
TERMINAL_PHASES = ("submitted", "expired")
# The timings are read from the DSH profile's own HV part, never repeated here.
_WINDOWS_MS = {w["afterPresentation"]: w["seconds"] * 1000
               for w in get_part("dsh", "listening", "hv_1").constraints["processingWindows"]}


class PlaybackError(GermanExamProfileError):
    """An event that the DSH two-play flow does not allow in the current state."""


def new_state(attempt_id: str) -> dict[str, Any]:
    if not isinstance(attempt_id, str) or not attempt_id:
        raise PlaybackError("attempt_id is required")
    return {"schemaVersion": SCHEMA_VERSION, "attemptId": attempt_id, "phase": "ready", "playsStarted": 0,
            "playsCompleted": 0, "windowDeadlineMs": None}


def _validate(state: dict[str, Any]) -> None:
    if state.get("schemaVersion") != SCHEMA_VERSION or state.get("phase") not in PHASES:
        raise PlaybackError("invalid playback state")
    if not isinstance(state.get("playsStarted"), int) or not 0 <= state["playsStarted"] <= MAX_PLAYS:
        raise PlaybackError("invalid play count")


def _tick(state: dict[str, Any], now_ms: int) -> dict[str, Any]:
    deadline = state["windowDeadlineMs"]
    if deadline is None or now_ms < deadline:
        return state
    if state["phase"] == "processing_1":
        return {**state, "phase": "awaiting_play_2", "windowDeadlineMs": None}
    if state["phase"] == "processing_2":
        return {**state, "phase": "expired", "windowDeadlineMs": None}
    return state


def advance(state: dict[str, Any], event: str, now_ms: int) -> dict[str, Any]:
    """Return the next state. Events: start_play, playback_ended, tick, submit."""
    _validate(state)
    if not isinstance(now_ms, int) or isinstance(now_ms, bool) or now_ms < 0:
        raise PlaybackError("now_ms must be a non-negative integer")
    state = _tick(state, now_ms)  # a lapsed window always resolves first, whatever the event is
    phase = state["phase"]
    if event == "tick":
        return state
    if event == "start_play":
        if phase == "ready":
            return {**state, "phase": "playing_1", "playsStarted": 1}
        if phase == "awaiting_play_2":
            return {**state, "phase": "playing_2", "playsStarted": 2}
        raise PlaybackError(f"cannot start a play in phase {phase!r} (plays started: {state['playsStarted']})")
    if event == "playback_ended":
        if phase in ("playing_1", "playing_2"):
            n = 1 if phase == "playing_1" else 2
            return {**state, "phase": f"processing_{n}", "playsCompleted": n, "windowDeadlineMs": now_ms + _WINDOWS_MS[n]}
        raise PlaybackError(f"no play is running in phase {phase!r}")
    if event == "submit":
        if phase == "processing_2":
            return {**state, "phase": "submitted", "windowDeadlineMs": None}
        raise PlaybackError(f"submission opens after the second presentation (phase {phase!r})")
    raise PlaybackError(f"unknown event {event!r}")


def recover(state: dict[str, Any], now_ms: int) -> dict[str, Any]:
    """Resume after a reload. A play that was interrupted counts as consumed: the play is over and
    its processing window starts now — it is never granted again."""
    _validate(state)
    if state["phase"] in ("playing_1", "playing_2"):
        state = advance(state, "playback_ended", now_ms)
    return advance(state, "tick", now_ms)


def can_start_play(state: dict[str, Any], now_ms: int) -> bool:
    try:
        advance(state, "start_play", now_ms)
        return True
    except PlaybackError:
        return False


def remaining_ms(state: dict[str, Any], now_ms: int) -> int | None:
    deadline = _tick(state, now_ms)["windowDeadlineMs"]
    return None if deadline is None else max(0, deadline - now_ms)
